In [1]:
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
DATA_DIR = Path("../data/raw/MachineLearningCSV/MachineLearningCVE")

csv_files = sorted(DATA_DIR.glob("*.csv"))

for file in csv_files:
    print(file.name)

Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
Friday-WorkingHours-Morning.pcap_ISCX.csv
Monday-WorkingHours.pcap_ISCX.csv
Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
Tuesday-WorkingHours.pcap_ISCX.csv
Wednesday-workingHours.pcap_ISCX.csv


In [3]:
Path.cwd()

WindowsPath('C:/CyberSOC-AI/notebooks')

In [4]:
list(DATA_DIR.iterdir())

[WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Friday-WorkingHours-Morning.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Monday-WorkingHours.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Tuesday-WorkingHours.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Wednesday-workingHours.pcap_ISCX.csv')]

In [5]:
print(DATA_DIR.resolve())
print(DATA_DIR.exists())
print(DATA_DIR.is_dir())
list(DATA_DIR.iterdir())

C:\CyberSOC-AI\data\raw\MachineLearningCSV\MachineLearningCVE
True
True


[WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Friday-WorkingHours-Morning.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Monday-WorkingHours.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Tuesday-WorkingHours.pcap_ISCX.csv'),
 WindowsPath('../data/raw/MachineLearningCSV/MachineLearningCVE/Wednesday-workingHours.pcap_ISCX.csv')]

In [6]:
for file in csv_files:
    sample = pd.read_csv(file, nrows=5)
    print(f"{file.name}: {sample.shape[1]} columnas")

Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv: 79 columnas
Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv: 79 columnas
Friday-WorkingHours-Morning.pcap_ISCX.csv: 79 columnas
Monday-WorkingHours.pcap_ISCX.csv: 79 columnas
Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv: 79 columnas
Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv: 79 columnas
Tuesday-WorkingHours.pcap_ISCX.csv: 79 columnas
Wednesday-workingHours.pcap_ISCX.csv: 79 columnas


In [7]:
sample = pd.read_csv(csv_files[0], nrows=5)

for i, column in enumerate(sample.columns, start=1):
    print(f"{i:02d}. {column}")

01.  Destination Port
02.  Flow Duration
03.  Total Fwd Packets
04.  Total Backward Packets
05. Total Length of Fwd Packets
06.  Total Length of Bwd Packets
07.  Fwd Packet Length Max
08.  Fwd Packet Length Min
09.  Fwd Packet Length Mean
10.  Fwd Packet Length Std
11. Bwd Packet Length Max
12.  Bwd Packet Length Min
13.  Bwd Packet Length Mean
14.  Bwd Packet Length Std
15. Flow Bytes/s
16.  Flow Packets/s
17.  Flow IAT Mean
18.  Flow IAT Std
19.  Flow IAT Max
20.  Flow IAT Min
21. Fwd IAT Total
22.  Fwd IAT Mean
23.  Fwd IAT Std
24.  Fwd IAT Max
25.  Fwd IAT Min
26. Bwd IAT Total
27.  Bwd IAT Mean
28.  Bwd IAT Std
29.  Bwd IAT Max
30.  Bwd IAT Min
31. Fwd PSH Flags
32.  Bwd PSH Flags
33.  Fwd URG Flags
34.  Bwd URG Flags
35.  Fwd Header Length
36.  Bwd Header Length
37. Fwd Packets/s
38.  Bwd Packets/s
39.  Min Packet Length
40.  Max Packet Length
41.  Packet Length Mean
42.  Packet Length Std
43.  Packet Length Variance
44. FIN Flag Count
45.  SYN Flag Count
46.  RST Flag Count
47. 

In [8]:
for file in csv_files:
    labels = pd.read_csv(file, usecols=[" Label"])
    
    print(f"\n{file.name}")
    print(f"Registros: {len(labels):,}")
    print(labels[" Label"].value_counts())


Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
Registros: 225,745
 Label
DDoS      128027
BENIGN     97718
Name: count, dtype: int64

Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
Registros: 286,467
 Label
PortScan    158930
BENIGN      127537
Name: count, dtype: int64

Friday-WorkingHours-Morning.pcap_ISCX.csv
Registros: 191,033
 Label
BENIGN    189067
Bot         1966
Name: count, dtype: int64

Monday-WorkingHours.pcap_ISCX.csv
Registros: 529,918
 Label
BENIGN    529918
Name: count, dtype: int64

Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
Registros: 288,602
 Label
BENIGN          288566
Infiltration        36
Name: count, dtype: int64

Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
Registros: 170,366
 Label
BENIGN                        168186
Web Attack � Brute Force        1507
Web Attack � XSS                 652
Web Attack � Sql Injection        21
Name: count, dtype: int64

Tuesday-WorkingHours.pcap_ISCX.csv
Registros: 445,909
 Label
BENIGN    

In [9]:
global_counts = pd.Series(dtype="int64")

for file in csv_files:
    labels = pd.read_csv(file, usecols=[" Label"])[" Label"]
    global_counts = global_counts.add(
        labels.value_counts(),
        fill_value=0
    )

global_counts = global_counts.astype("int64").sort_values(ascending=False)

print(f"Total de registros: {global_counts.sum():,}")
print(f"Número de clases: {len(global_counts)}")
print()
print(global_counts)

Total de registros: 2,830,743
Número de clases: 15

 Label
BENIGN                        2273097
DoS Hulk                       231073
PortScan                       158930
DDoS                           128027
DoS GoldenEye                   10293
FTP-Patator                      7938
SSH-Patator                      5897
DoS slowloris                    5796
DoS Slowhttptest                 5499
Bot                              1966
Web Attack � Brute Force         1507
Web Attack � XSS                  652
Infiltration                       36
Web Attack � Sql Injection         21
Heartbleed                         11
dtype: int64


## Planteamiento del problema

El dataset CIC-IDS2017 contiene 2.830.743 registros distribuidos en 15 clases. 
El tráfico benigno representa, aproximadamente, el 80 % de los registros, mientras 
que el resto corresponde a diferentes tipos de ataques.

CyberSOC AI plantea el problema principal como una **detección binaria de anomalías**:

- **Normal:** tráfico etiquetado como `BENIGN`.
- **Anómalo:** tráfico correspondiente a cualquier categoría de ataque.

Las etiquetas originales de los ataques se conservarán para realizar posteriormente 
un análisis detallado del comportamiento de los modelos frente a diferentes tipos 
de amenaza.

### Hipótesis de trabajo

Se evaluará si un modelo de Deep Learning basado en autoencoder, entrenado para 
aprender el comportamiento del tráfico benigno, permite detectar tráfico anómalo 
mediante el error de reconstrucción y obtener un rendimiento competitivo frente 
a un modelo clásico supervisado de referencia.

### Criterios de evaluación

Debido al desbalance existente entre las clases, la accuracy no se 
utilizará de forma aislada. La evaluación considerará métricas como precisión, 
recall, F1-score, ROC-AUC, tasa de falsos positivos y matriz de confusión.

También se analizará el comportamiento frente a distintos tipos de ataque cuando 
la cantidad de muestras permita obtener conclusiones significativas.

In [10]:
sample = pd.read_csv(csv_files[0], nrows=5)

print("ANTES:")
print(repr(sample.columns[0]))
print(repr(sample.columns[-1]))

sample.columns = sample.columns.str.strip()

print("\nDESPUÉS:")
print(repr(sample.columns[0]))
print(repr(sample.columns[-1]))

ANTES:
' Destination Port'
' Label'

DESPUÉS:
'Destination Port'
'Label'


In [11]:
for file in csv_files:
    df_temp = pd.read_csv(file)
    df_temp.columns = df_temp.columns.str.strip()

    # Solo variables numéricas
    numeric = df_temp.select_dtypes(include=[np.number])

    null_count = df_temp.isnull().sum().sum()
    inf_count = np.isinf(numeric).sum().sum()

    print(f"{file.name}")
    print(f"  Registros: {len(df_temp):,}")
    print(f"  Valores nulos: {null_count:,}")
    print(f"  Valores infinitos: {inf_count:,}")
    print()

Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
  Registros: 225,745
  Valores nulos: 4
  Valores infinitos: 64

Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
  Registros: 286,467
  Valores nulos: 15
  Valores infinitos: 727

Friday-WorkingHours-Morning.pcap_ISCX.csv
  Registros: 191,033
  Valores nulos: 28
  Valores infinitos: 216

Monday-WorkingHours.pcap_ISCX.csv
  Registros: 529,918
  Valores nulos: 64
  Valores infinitos: 810

Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
  Registros: 288,602
  Valores nulos: 18
  Valores infinitos: 396

Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
  Registros: 170,366
  Valores nulos: 20
  Valores infinitos: 250

Tuesday-WorkingHours.pcap_ISCX.csv
  Registros: 445,909
  Valores nulos: 201
  Valores infinitos: 327

Wednesday-workingHours.pcap_ISCX.csv
  Registros: 692,703
  Valores nulos: 1,008
  Valores infinitos: 1,586



In [12]:
problem_columns = {}
total_problem_rows = 0

for file in csv_files:
    df_temp = pd.read_csv(file)
    df_temp.columns = df_temp.columns.str.strip()

    numeric = df_temp.select_dtypes(include=[np.number])

    # Filas que contienen al menos un NaN o infinito
    problem_mask = (
        df_temp.isnull().any(axis=1) |
        np.isinf(numeric).any(axis=1)
    )

    problem_rows = problem_mask.sum()
    total_problem_rows += problem_rows

    # Problemas por columna
    nulls = df_temp.isnull().sum()
    infs = pd.Series(
        np.isinf(numeric).sum(),
        index=numeric.columns
    )

    columns_with_problems = (
        nulls[nulls > 0]
        .add(infs[infs > 0], fill_value=0)
        .sort_values(ascending=False)
    )

    print(f"\n{file.name}")
    print(f"Filas con algún problema: {problem_rows:,}")
    print(columns_with_problems)

print(f"\nTOTAL filas con algún NaN o infinito: {total_problem_rows:,}")


Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
Filas con algún problema: 34
Flow Bytes/s      34.0
Flow Packets/s    34.0
dtype: float64

Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
Filas con algún problema: 371
Flow Bytes/s      371.0
Flow Packets/s    371.0
dtype: float64

Friday-WorkingHours-Morning.pcap_ISCX.csv
Filas con algún problema: 122
Flow Bytes/s      122.0
Flow Packets/s    122.0
dtype: float64

Monday-WorkingHours.pcap_ISCX.csv
Filas con algún problema: 437
Flow Bytes/s      437.0
Flow Packets/s    437.0
dtype: float64

Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
Filas con algún problema: 207
Flow Bytes/s      207.0
Flow Packets/s    207.0
dtype: float64

Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
Filas con algún problema: 135
Flow Bytes/s      135.0
Flow Packets/s    135.0
dtype: float64

Tuesday-WorkingHours.pcap_ISCX.csv
Filas con algún problema: 264
Flow Bytes/s      264.0
Flow Packets/s    264.0
dtype: float64

Wednesday-workin

### Tratamiento de valores no finitos

El análisis de calidad identificó 2.867 registros que contienen al menos un valor
nulo (`NaN`) o infinito, sobre un total de 2.830.743 registros (aproximadamente
el 0,10 % del conjunto de datos).

Los valores problemáticos se concentran en las variables `Flow Bytes/s` y
`Flow Packets/s`.

Debido a la reducida proporción de registros afectados, se ha decidido eliminar
estas observaciones durante el preprocesamiento en lugar de aplicar técnicas de
imputación. De esta forma se evita introducir valores artificiales en variables
relacionadas directamente con el comportamiento del tráfico de red.

In [13]:
total_duplicates = 0

for file in csv_files:
    df_temp = pd.read_csv(file)
    df_temp.columns = df_temp.columns.str.strip()

    duplicates = df_temp.duplicated().sum()
    total_duplicates += duplicates

    print(f"{file.name}")
    print(f"  Filas duplicadas: {duplicates:,}")
    print()

print(f"TOTAL duplicados dentro de los ficheros: {total_duplicates:,}")

Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
  Filas duplicadas: 2,633

Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
  Filas duplicadas: 72,353

Friday-WorkingHours-Morning.pcap_ISCX.csv
  Filas duplicadas: 6,888

Monday-WorkingHours.pcap_ISCX.csv
  Filas duplicadas: 26,935

Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
  Filas duplicadas: 35,630

Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
  Filas duplicadas: 6,066

Tuesday-WorkingHours.pcap_ISCX.csv
  Filas duplicadas: 24,065

Wednesday-workingHours.pcap_ISCX.csv
  Filas duplicadas: 81,909

TOTAL duplicados dentro de los ficheros: 256,479


In [14]:
duplicate_counts_by_label = pd.Series(dtype="int64")

for file in csv_files:
    df_temp = pd.read_csv(file)
    df_temp.columns = df_temp.columns.str.strip()

    # Seleccionamos las filas consideradas duplicadas
    duplicated_rows = df_temp[df_temp.duplicated()]

    counts = duplicated_rows["Label"].value_counts()

    duplicate_counts_by_label = duplicate_counts_by_label.add(
        counts,
        fill_value=0
    )

duplicate_counts_by_label = (
    duplicate_counts_by_label
    .astype("int64")
    .sort_values(ascending=False)
)

print("Duplicados por clase:")
print(duplicate_counts_by_label)

print(f"\nTOTAL: {duplicate_counts_by_label.sum():,}")

Duplicados por clase:
Label
BENIGN                      124711
PortScan                     68111
DoS Hulk                     58224
SSH-Patator                   2678
FTP-Patator                   2005
DoS slowloris                  411
DoS Slowhttptest               271
Web Attack � Brute Force        37
Bot                             13
DDoS                            11
DoS GoldenEye                    7
dtype: int64

TOTAL: 256,479


### Tratamiento de registros duplicados

El análisis ha identificado 256.479 registros duplicados dentro de los distintos
ficheros del dataset, lo que representa aproximadamente un 9,1 % de los
2.830.743 registros originales.

Los duplicados no se concentran únicamente en el tráfico benigno, sino que
también afectan a determinadas categorías de ataque,
especialmente PortScan y DoS Hulk.

Se decide eliminar los registros duplicados antes de realizar la división
entre los conjuntos de entrenamiento, validación y prueba. Esta decisión busca
reducir la redundancia del conjunto de datos y evitar que
observaciones idénticas puedan aparecer simultáneamente en entrenamiento y
prueba, lo que podría producir una estimación muy optimista del
rendimiento de los modelos.

Durante la construcción del conjunto procesado se comprobarán también posibles
duplicados existentes entre ficheros diferentes.

In [15]:
sample = pd.read_csv(csv_files[0], nrows=1000)
sample.columns = sample.columns.str.strip()

print("Tipos de columnas:")
print(sample.dtypes.value_counts())

print("\nColumnas no numéricas:")
print(sample.select_dtypes(exclude=[np.number]).columns.tolist())

Tipos de columnas:
int64      54
float64    24
str         1
Name: count, dtype: int64

Columnas no numéricas:
['Label']


In [16]:
reference_columns = None

for file in csv_files:
    df_sample = pd.read_csv(file, nrows=1000)
    df_sample.columns = df_sample.columns.str.strip()

    numeric_columns = df_sample.select_dtypes(include=[np.number]).columns
    non_numeric_columns = df_sample.select_dtypes(exclude=[np.number]).columns

    if reference_columns is None:
        reference_columns = df_sample.columns.tolist()

    same_columns = df_sample.columns.tolist() == reference_columns

    print(file.name)
    print(f"  Columnas totales: {len(df_sample.columns)}")
    print(f"  Numéricas: {len(numeric_columns)}")
    print(f"  No numéricas: {non_numeric_columns.tolist()}")
    print(f"  Mismo esquema: {same_columns}")
    print()

Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
  Columnas totales: 79
  Numéricas: 78
  No numéricas: ['Label']
  Mismo esquema: True

Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
  Columnas totales: 79
  Numéricas: 78
  No numéricas: ['Label']
  Mismo esquema: True

Friday-WorkingHours-Morning.pcap_ISCX.csv
  Columnas totales: 79
  Numéricas: 78
  No numéricas: ['Label']
  Mismo esquema: True

Monday-WorkingHours.pcap_ISCX.csv
  Columnas totales: 79
  Numéricas: 78
  No numéricas: ['Label']
  Mismo esquema: True

Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
  Columnas totales: 79
  Numéricas: 78
  No numéricas: ['Label']
  Mismo esquema: True

Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
  Columnas totales: 79
  Numéricas: 78
  No numéricas: ['Label']
  Mismo esquema: True

Tuesday-WorkingHours.pcap_ISCX.csv
  Columnas totales: 79
  Numéricas: 78
  No numéricas: ['Label']
  Mismo esquema: True

Wednesday-workingHours.pcap_ISCX.csv
  Columnas totales:

In [17]:
PROCESSED_DIR = Path("../data/processed")
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

print(PROCESSED_DIR.resolve())

C:\CyberSOC-AI\data\processed


In [18]:
dataframes = []

for file in csv_files:
    print(f"Procesando: {file.name}")

    df_temp = pd.read_csv(file)

    #Limpiar nombres de columnas
    df_temp.columns = df_temp.columns.str.strip()

    #Sustituir infinitos por NaN
    df_temp.replace([np.inf, -np.inf], np.nan, inplace=True)

    #Eliminar registros con valores no válidos
    df_temp.dropna(inplace=True)

    dataframes.append(df_temp)

print("\nUniendo ficheros...")

df = pd.concat(dataframes, ignore_index=True)

print(f"Registros después de eliminar NaN/inf: {len(df):,}")

Procesando: Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
Procesando: Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
Procesando: Friday-WorkingHours-Morning.pcap_ISCX.csv
Procesando: Monday-WorkingHours.pcap_ISCX.csv
Procesando: Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
Procesando: Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
Procesando: Tuesday-WorkingHours.pcap_ISCX.csv
Procesando: Wednesday-workingHours.pcap_ISCX.csv

Uniendo ficheros...
Registros después de eliminar NaN/inf: 2,827,876


In [19]:
before_duplicates = len(df)

df = df.drop_duplicates().reset_index(drop=True)

after_duplicates = len(df)
removed_duplicates = before_duplicates - after_duplicates

print(f"Registros antes de eliminar duplicados: {before_duplicates:,}")
print(f"Registros después de eliminar duplicados: {after_duplicates:,}")
print(f"Duplicados eliminados globalmente: {removed_duplicates:,}")

Registros antes de eliminar duplicados: 2,827,876
Registros después de eliminar duplicados: 2,520,798
Duplicados eliminados globalmente: 307,078


In [20]:
df["is_attack"] = (df["Label"] != "BENIGN").astype(int)

print(df[["Label", "is_attack"]].head())
print()
print(df["is_attack"].value_counts())

    Label  is_attack
0  BENIGN          0
1  BENIGN          0
2  BENIGN          0
3  BENIGN          0
4  BENIGN          0

is_attack
0    2095057
1     425741
Name: count, dtype: int64


In [21]:
%pip install pyarrow

Note: you may need to restart the kernel to use updated packages.


In [22]:
OUTPUT_FILE = PROCESSED_DIR / "cic_ids2017_clean.parquet"

df.to_parquet(OUTPUT_FILE, index=False)

print(f"Dataset guardado en: {OUTPUT_FILE.resolve()}")
print(f"Registros: {len(df):,}")
print(f"Columnas: {len(df.columns)}")

Dataset guardado en: C:\CyberSOC-AI\data\processed\cic_ids2017_clean.parquet
Registros: 2,520,798
Columnas: 80
